src/hp/derp/http/server.rs

```diff
@@ -109,145 +463,199 @@ where
+    /// Serve the tls connection
+    async fn tls_serve_connection(
+        self,
+        stream: TcpStream,
+        acceptor: TlsAcceptor,
+        rustls_config: Arc<rustls::ServerConfig>,
+    ) -> Result<()> {
+        match acceptor {
+            TlsAcceptor::LetsEncrypt(a) => match a.accept(stream).await? {
+                None => {
+                    info!("received TLS-ALPN-01 validation request");
+                }
+                Some(start_handshake) => {
+                    let tls_stream = start_handshake.into_stream(rustls_config).await?;
+                    self.serve_connection(MaybeTlsStreamServer::Tls(tls_stream))
+                        .await?;
+                }
+            },
+            TlsAcceptor::Manual(a) => {
+                let tls_stream = a.accept(stream).await?;
+                self.serve_connection(MaybeTlsStreamServer::Tls(tls_stream))
+                    .await?;
+            }
+        }
+        Ok(())
+    }
 
-        let listener = TcpListener::bind(&addr).await?;
-        // We need the assigned address for the client to send it messages.
-        let addr = listener.local_addr()?;
+    /// Wrapper for the actual http connection (with upgrades)
+    async fn serve_connection<I>(self, io: I) -> Result<()>
+    where
+        I: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send + Sync + 'static,
+    {
+        Http::new()
+            .serve_connection(io, self)
+            .with_upgrades()
+            .await?;
+        Ok(())
+    }
+}
 
-        // For this example, a oneshot is used to signal that after 1 request,
-        // the server should be shutdown.
-        let (tx, mut rx) = oneshot::channel::<()>();
+#[derive(Default)]
+struct Handlers(pub HashMap<(Method, &'static str), HyperFn>);
 
-        // Spawn server on the default executor,
-        // which is usually a thread-pool from tokio default runtime.
-        let server_task = tokio::task::spawn(async move {
-            loop {
-                tokio::select! {
-                    biased;
-                    _ = &mut rx => {
-                        return Ok::<_, anyhow::Error>(());
-                    }
-
-                    conn = listener.accept() => {
-                        let (stream, _) = conn?;
-                        let derp_client_handler = derp_client_handler.clone();
-                        tokio::task::spawn(async move {
-                            if let Err(err) = Http::new()
-                                .serve_connection(stream, derp_client_handler)
-                                .with_upgrades()
-                                .await
-                            {
-                                eprintln!("Failed to serve connection: {:?}", err);
-                            }
-                        });
-                    }
-                }
-            }
+impl std::fmt::Debug for Handlers {
```

Comment:
derive_more?

Reply:
Okay, this was the only one I couldn't figure out how to bend to my will using `derive_more`